lazy_dataset/core.py

```diff
@@ -2313,6 +2397,95 @@ def __iter__(self):
             print(f'Dropped {dropped_count} examples')
 
 
+class CacheDataset(Dataset):
+    def __init__(self, dataset: Dataset, keep_mem_free=None,
+                 immutable_warranty: str = 'pickle') -> None:
+        super().__init__()
+        assert dataset.indexable, (
+            'Cache Dataset only works if dataset is indexable!'
+        )
+        self.dataset = dataset
+        self.cache = {}
+
+        self._keep_mem_free = self._get_memory_size(keep_mem_free)
+        self.immutable_warranty = immutable_warranty
+
+        if immutable_warranty == 'pickle':
+            self._serialize = pickle.dumps
+            self._deserialize = pickle.loads
+        elif immutable_warranty == 'copy':
+            self._serialize = lambda x: x
+            self._deserialize = deepcopy
+        else:
+            raise ValueError(immutable_warranty)
+
+    @property
+    def indexable(self) -> bool:
+        return True
+
+    def keys(self) -> list:
+        return self.dataset.keys()
+
+    @staticmethod
+    def _get_memory_size(keep_mem_free):
+        if keep_mem_free is None:
+            return None
+
+        assert isinstance(keep_mem_free, (list, tuple)) \
+               and len(keep_mem_free) == 2, (
+            f'keep_mem_free has to be a tuple of value and unit, and not '
+            f'{keep_mem_free}'
+        )
+
+        value, unit = keep_mem_free
+
+        import psutil
+
+        if unit == 'GB':
+            return value * 1024 ** 3
+        elif unit == 'fraction':
+            assert 0 <= value < 1, value
+            return psutil.virtual_memory().total * value
+        else:
+            raise ValueError(f'Unknown unit for memory size: {unit}')
+
+    def __getitem__(self, item):
+        if item not in self.cache:
```

Comment:
Can you handle the different types of item? e.g.:
```python
        if isinstance(item, numbers.Integral):
            item = self.keys()[item]
        if isinstance(item, str):
            ...  # do caching
        else:
            return super().__getitem__(item)
```

Reply:
Done